src/hashsum/digest.rs

```diff
@@ -0,0 +1,266 @@
+extern crate digest;
+extern crate md5;
+extern crate rustc_serialize;
+extern crate sha1;
+extern crate sha2;
+extern crate sha3;
+
+use digest::digest::{Input, ExtendableOutput, XofReader};
+
+pub trait Digest {
+    fn new() -> Self where Self: Sized;
+    fn input(&mut self, input: &[u8]);
+    fn result(&mut self, out: &mut [u8]);
+    fn reset(&mut self);
+    fn output_bits(&self) -> usize;
+    fn output_bytes(&self) -> usize {
+        (self.output_bits() + 7) / 8
+    }
+    fn result_str(&mut self) -> String {
+        use self::rustc_serialize::hex::ToHex;
```

Comment:
I didn't notice this before, but we should also probably switch to the `hex` crate for this, as `rustc_serialize` is deprecated.